sdk/python/v1beta1/kubeflow/katib/api/katib_client.py

```diff
@@ -411,9 +450,173 @@ def tune(
+                    p_value.name = p_name
+                    experiment_params.append(p_value)
+                    trial_params.append(models.V1beta1TrialParameterSpec(name=p_name, reference=p_name))
+                elif p_value is not None:
+                    old_attr = getattr(lora_config, p_name, None)
+                    if old_attr is not None:
+                        value = type(old_attr)(p_value)
+                    setattr(lora_config, p_name, value)
+
+            # create init container spec.
+            init_container_spec = client.V1Container(
+                name=constants.STORAGE_INITIALIZER,
+                image=constants.STORAGE_INITIALIZER_IMAGE,
+                args=[
+                    "--model_provider",
+                    mp,
+                    "--model_provider_parameters",
+                    json.dumps(model_provider_parameters.__dict__, cls=utils.SetEncoder), 
+                    "--dataset_provider",
+                    dp,
+                    "--dataset_provider_parameters",
+                    json.dumps(dataset_provider_parameters.__dict__),
+                ],
+                volume_mounts=[constants.STORAGE_INITIALIZER_VOLUME_MOUNT],
+            )
+
+            lora_config = json.dumps(lora_config.__dict__, cls=utils.SetEncoder)
+            training_args = json.dumps(training_args.to_dict())
+            # create app container spec.
+            container_spec = client.V1Container(
+                name=constants.DEFAULT_PRIMARY_CONTAINER_NAME,
+                image=constants.TRAINER_TRANSFORMER_IMAGE,
+                args=[
+                    "--model_uri",
+                    model_provider_parameters.model_uri,
+                    "--transformer_type",
+                    model_provider_parameters.transformer_type.__name__,
+                    "--model_dir",
+                    constants.VOLUME_PATH_MODEL,
+                    "--dataset_dir",
+                    constants.VOLUME_PATH_DATASET,
+                    "--lora_config",
+                    f"'{lora_config}'",
+                    "--training_parameters",
+                    f"'{training_args}'",
+                ],
+                volume_mounts=[constants.STORAGE_INITIALIZER_VOLUME_MOUNT],
+                env=env if env else None,
+                env_from=env_from if env_from else None,
+                resources=resources_per_trial,
+            )
+
+            pod_spec = client.V1PodTemplateSpec(
+                metadata=models.V1ObjectMeta(
+                    annotations={"sidecar.istio.io/inject": "false"}
+                ),
+                spec=client.V1PodSpec(
+                    restart_policy="Never",
+                    containers=[container_spec],
+                    init_containers=[init_container_spec],
+                    volumes=[constants.STORAGE_INITIALIZER_VOLUME],
+                ),
+            )
+
+        # Create Trial specification.
+        trial_spec = client.V1Job(
```

Comment:
This should be done only when `tune` is used with objective.

Reply:
@andreyvelich Sorry, I didn't understand the difference. Since the `V1beta1TrialTemplate` requires a `trial_spec`, if we only provide this when `tune` is used with an objective, what should we do when `tune` is used with external models and datasets?

Reply:
>what should we do when tune is used with external models and datasets?

In that case, Trial spec is a PyTorchJob, like in this example: https://github.com/kubeflow/katib/blob/master/examples/v1beta1/kubeflow-training-operator/pytorchjob-mnist.yaml

Reply:
@andreyvelich I understand. I am considering adding PyTorchJob as an additional option for setting `resources_per_trial` for both the custom objective function and external models and datasets. I plan to implement the PyTorchJob functionality in a separate PR. How does that sound?

Reply:
Yes, let's do that after we support LLM trainer within `tune` API.